sway_libs/src/merkle_proof.sw

```diff
@@ -0,0 +1,151 @@
+// TODO: Function definitions that use arrays should be updated to Vecs once
+// https://github.com/FuelLabs/fuels-rs/issues/353 is resolved
+
+library merkle_proof;
+
+use std::{hash::sha256, option::Option, revert::{require, revert}, vec::Vec};
+
+pub enum ProofError {
+    InvalidMultiProof: (),
+}
+
+fn hash_pair(a: b256, b: b256) -> b256 {
+    if a <= b {
+        // Hash(a + b)
+        sha256((a, b))
+    } else {
+        // Hash(b + a)
+        sha256((b, a))
+    }
+}
+
+/// This function will compute a merkle root using the multiple merkle leaves and proof given.
+///
+/// # Arguments
+///
+/// * `merkle_leaves` - The hashes of relevant leaves for the merkle proof.
+/// * `proof` - The merkle proof that will be used to traverse the merkle tree and compute a root.
+/// * 'proof_flags' - The flags used to determine which hashes to use in order to compute the merkle root.
+///
+/// # Reverts
+///
+/// * When an incorrect number of proof flags for the multi-proof is given.
+pub fn process_multi_proof(merkle_leaves: [b256;
+2], proof: b256, proof_flags: [bool;
+2]) -> b256 {
+    // TODO: These should not be hard-coded. They are only to be used as placeholders until
+    // https://github.com/FuelLabs/fuels-rs/issues/353 is resolved
+    let total_hashes = 2;
+    let merkle_leaves_len = 2;
+    let proof_len = 1;
+    // let total_hashes = proof_flags.len();
+    // let merkle_leaves_len = merkle_leaves.len();
+    // let proof_len = proof.len();
+    require(merkle_leaves_len + proof_len - 1 == total_hashes, ProofError::InvalidMultiProof);
+
+    let mut hashes: Vec<b256> = ~Vec::new();
+    let mut iterator = 0;
+    let mut leaf_pos = 0;
+    let mut hash_pos = 0;
+    let mut proof_pos = 0;
+
+    // For each step we find a suitable `a` and `b` to hash:
+    // `a` - A leaf hash if there are unused leaves or a computed hash
+    // `b` - Based on the provided proof flags:
+    //      * True - A leaf hash if there are unused leaves or a computed hash
+    //      * False - A provided proof hash
+    while iterator < total_hashes {
+        let a = if leaf_pos < merkle_leaves_len {
+            leaf_pos += 1;
+            merkle_leaves[leaf_pos - 1]
+        } else {
+            hash_pos += 1;
+            hashes.get(hash_pos - 1).unwrap()
+        };
+
+        let b = if proof_flags[iterator] {
+            if leaf_pos < merkle_leaves_len {
```

Comment:
You could consider adding a helper function such as get_hash(...) which returns either a leaf hash or a computed hash.  Although you would only use it twice so it may be more trouble than its worth.